kotlin/workflow-core/src/main/java/com/squareup/workflow/StatefulWorkflow.kt

```diff
@@ -0,0 +1,136 @@
+ * the outside world, and be [saved][snapshotState] to a serialized form to be
+ * [restored][restoreState] later.
+ *
+ * The basic purpose of a `Workflow` is to take some [input][InputT] and return a
+ * [rendering][RenderingT]. To that end, a workflow may keep track of internal [state][StateT],
+ * recursively ask other workflows to render themselves, subscribe to data streams from the outside
+ * world, and handle events both from its [renderings][WorkflowContext.onEvent] and from workflows
+ * it's delegated to (its "children"). A `Workflow` may also emit [output events][OutputT] up to its
+ * parent `Workflow`.
+ *
+ * Workflows form a tree, where each workflow can have zero or more child workflows. Child workflows
+ * are started as necessary whenever another workflow asks for them, and are cleaned up automatically
+ * when they're no longer needed. [Input][InputT] propagates down the tree, [outputs][OutputT] and
+ * [renderings][RenderingT] propagate up the tree.
+ *
+ * @param InputT Typically a data class that is used to pass configuration information or bits of
+ * state that the workflow can always get from its parent and needn't duplicate in its own state.
+ * May be [Unit] if the workflow does not need any input data.
+ *
+ * @param StateT Typically a data class that contains all of the internal state for this workflow.
+ * The state is seeded via [input][InputT] in [initialState]. It can be [serialized][snapshotState]
+ * and later used to [restore][restoreState] the workflow. **Implementations of the `Workflow`
+ * interface should not generally contain their own state directly.** They may inject objects like
+ * instances of their child workflows, or network clients, but should not contain directly mutable
+ * state. This is the only type parameter that a parent workflow needn't care about for its children,
+ * and may just use star (`*`) instead of specifying it. May be [Unit] if the workflow does not have
+ * any internal state (see [StatelessWorkflow]).
+ *
+ * @param OutputT Typically a sealed class that represents "events" that this workflow can send
+ * to its parent.
+ * May be [Nothing] if the workflow doesn't need to emit anything.
+ *
+ * @param RenderingT The value returned to this workflow's parent during [composition][compose].
+ * Typically represents a "view" of this workflow's input, current state, and children's renderings.
+ * A workflow that represents a UI component may use a view model as its rendering type.
+ *
+ * @see StatelessWorkflow
+ */
+abstract class StatefulWorkflow<
+    in InputT : Any,
+    StateT : Any,
+    out OutputT : Any,
+    out RenderingT : Any
+    > : Workflow<InputT, OutputT, RenderingT> {
+
+  /**
+   * Called when the state machine is first started to get the initial state.
+   */
+  abstract fun initialState(input: InputT): StateT
+
+  /**
+   * Called whenever [WorkflowContext.compose] is about to get a new [input][InputT] value, to allow
```

Comment:
Called from `WorkflowContext.compose` instead of `initialState` when a workflow is already running. This 
allows the workflow to detect changes in input, and possibly change its state in response.